client/consensus/common/src/level_monitor.rs

```diff
@@ -0,0 +1,269 @@
+			let route = sp_blockchain::tree_route(self.backend.blockchain(), finalized, leaf)
+				.expect("Route from finalized to leaf should be available; qed");
+			if !route.retracted().is_empty() {
+				continue
+			}
+			std::iter::once(route.common_block()).chain(route.enacted()).for_each(|elem| {
+				if !self.freshness.contains_key(&elem.hash) {
+					// Use the block height value as the freshness.
+					self.import_counter = elem.number.unique_saturated_into();
+					self.update(elem.number, elem.hash);
+				}
+			});
+			if self.import_counter > counter_max {
+				counter_max = self.import_counter;
+			}
+		}
+
+		self.import_counter = counter_max;
+		self.lowest_level = self.backend.blockchain().info().finalized_number;
+	}
+
+	/// Check and enforce the limit bound at the given height.
+	///
+	/// In practice this will enforce the given height in having a number of blocks less than
+	/// the limit passed to the constructor.
+	///
+	/// If the given level is found to have a number of blocks greater than or equal the limit
+	/// then the limit is enforced by chosing one (or more) blocks to remove.
+	/// The removal strategy is driven by the block freshness.
+	///
+	/// A block freshness is determined by the most recent leaf freshness descending from the block
+	/// itself. In other words its freshness is equal to its more "fresh" descendant.
+	///
+	/// The least "fresh" blocks are eventually removed.
+	pub fn enforce_limit(&mut self, number: NumberFor<Block>) {
+		if self.import_counter == 0 {
+			// First call detected, restore using the backend persistent information.
+			self.restore();
+		}
+
+		let level = self.levels.entry(number).or_default();
+		if level.len() < self.level_limit {
+			return
+		}
+
+		log::debug!(
+			target: "parachain",
+			"Detected leaves overflow at height {}, removing obsolete blocks", number
+		);
+
+		let mut leaves = self.backend.blockchain().leaves().unwrap_or_default();
+		// Sort leaves by freshness (most recent first).
+		leaves.sort_unstable_by(|a, b| self.freshness.get(a).cmp(&self.freshness.get(b)));
+
+		// This may not be the most efficient way to remove **multiple** entries, but is the clear
+		// one. Should be considered that in normal conditions the number of blocks to remove is 0
+		// or 1, it is not worth to complicate the code too much.
+		let remove_count = level.len() - self.level_limit + 1;
+		(0..remove_count).for_each(|_| self.remove_block(number, &leaves));
+	}
+
+	/// This will search for the less fresh block and removes it along with all its descendants.
+	/// Leaves should have already been ordered by "freshness" (older first).
+	fn remove_block(&mut self, number: NumberFor<Block>, leaves: &[Block::Hash]) {
+		let mut candidate_freshest_leaf_idx = usize::MAX;
+		let mut candidate_freshest_route = None;
+
+		let blockchain = self.backend.blockchain();
+		let best_hash = blockchain.info().best_hash;
+
+		for blk_hash in self.levels.entry(number).or_default().iter() {
```

Comment:
Shouldn't we also not start here with the oldest block? If we change `levels` to `HashMap<Hash, Vec<>>` we could get the ordering oldest to freshest for free.

Reply:
This is true. If you use `HashMap<Hash, Vec<>>` you already find the entries within the vector sorted by arrival order.

But the point of my overall strategy is not to remove the older element at a given level. But to remove the one according to the search strategy I've described in my other comments.

So having it already sorted is not super helpful.

I went for an `HashMap<Hash, HashSet<>>` because the HashSet is faster for removal of elements at random indices.